cmd/swarm/swarm-smoke/sliding_window.go

```diff
@@ -0,0 +1,128 @@
+// Copyright 2018 The go-ethereum Authors
+// This file is part of go-ethereum.
+//
+// go-ethereum is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// go-ethereum is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with go-ethereum. If not, see <http://www.gnu.org/licenses/>.
+
+package main
+
+import (
+	"bytes"
+	"fmt"
+	"math/rand"
+	"sync"
+	"time"
+
+	"github.com/ethereum/go-ethereum/log"
+	"github.com/ethereum/go-ethereum/metrics"
+	"github.com/ethereum/go-ethereum/swarm/testutil"
+	"github.com/pborman/uuid"
+
+	cli "gopkg.in/urfave/cli.v1"
+)
+
+type uploadResult struct {
+	hash   string
+	digest []byte
+}
+
+func slidingWindow(c *cli.Context) error {
+	// test dscription:
+	// 1. upload repeatedly the same file size, maintain a slice in which swarm hashes are stored, first hash at idx=0
+	// 2. select a random node, start downloading the hashes, starting with the LAST one first (it should always be availble), till the FIRST hash
+	// 3. when
+
+	defer func(now time.Time) {
+		totalTime := time.Since(now)
+
+		log.Info("total time", "time", totalTime)
+		metrics.GetOrRegisterCounter("sliding-window.total-time", nil).Inc(int64(totalTime))
+	}(time.Now())
+
+	generateEndpoints(scheme, cluster, appName, from, to)
+	storeSize = storeSize * 4096 //store size is in chunks - transform to bytes
+	hashes := []uploadResult{}   //swarm hashes of the uploads
+	nodes := to - from
+	networkCapacity := float64(storeSize) * float64(nodes)
+	const iterationTimeout = 30 * time.Second
+	log.Info("sliding window test started", "store size(kb)", int(storeSize/1000), "nodes", nodes, "filesize(kb)", filesize, "network capacity(kb)", int(networkCapacity/1000), "timeout", timeout)
+	uploadedBytes := 0
+	networkDepth := 0
+	errored := false
+outer:
+	for {
+		seed := int(time.Now().UnixNano() / 1e6)
+		log.Info("uploading to "+endpoints[0]+" and syncing", "seed", seed)
+
+		randomBytes := testutil.RandomBytes(seed, filesize*1000)
```

Comment:
Could this be avoided? We are generating in memory bytes that can be up to gigabytes. A simple limited reader from rand.Read would avoid so large allocation. Something like 

```
r = io.LimitReader(rand.Reader, filesize*1000)
```

and use TeeReader to get the hash

```
h := md5.New()
r := io.TeeReader(io.LimitReader(rand.Reader, filesize*1000), h)
```

This way, there is no large byte slice allocation.